students/guillaume/Q2_session02/integration.py

```diff
@@ -40,6 +40,77 @@ def sectors(*args):
     return values, bands
 
 
+def trapz_p(func, a, b, precision=10, **func_args):
+    '''
+    This function uses a more generic approach to trapeziodal numerical
+    integration in order to allow non constant size of the sub integration
+    sectors hence it is less efficient but more versatile
+    '''
+    # Test order of a & b
+    if a > b:
+        a, b = b, a
+
+    def loc_f(x):
+        nonlocal func_args
+        return func(x, **func_args)
+
+    x_l, f_x_l = sectors_precisions(loc_f, a, b, precision)
+
+    def band_f(lst, func):
+        l_v = len(lst)
+        return [func(x1, x2)
+                for x1, x2 in zip(islice(lst, 0, l_v - 1),
+                                  islice(lst, 1, l_v))]
+
+    band = band_f(x_l, lambda x1, x2: x2 - x1)
+    f_band = band_f(f_x_l, lambda x1, x2: 0.5 * (x1 + x2))
+    integral = sum([x * y for x, y in zip(band, f_band)])
+
+    return integral
+
+
+def sectors_precisions(*args):
+    '''
+    Assuming that f meets Riemann criterions for integration
+    default size of an integration sub domain is (1/20 x |b - a|)
+    for a sub dommain [x1, x2], where x2 >x 1
+    Z = |f(x2) - f(x1)| will be compared to a precision value
+    if Z bigger than precision x2 size will be decrease until the inequation
+    is accurate.
+    There is a test to avoid that calculations would run for too long
+    '''
+    func, a, b, precision = args
```

Comment:
why `*args`, rather than putting `func, a, b, precision` in the function signature?